libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -816,7 +816,7 @@ export class QnAMakerDialog extends WaterfallDialog implements QnAMakerDialogCon
 
         // V4 API logic
         // If the hostname contains all the necessary information, return it
-        if (/^https:\/\/.*\.azurewebsites\.net\/qnamaker\/?/i.test(host)) {
+        if (/^https:\/\/.*\..*\.\w{2,}\/qnamaker\/?/i.test(host)) {
```

Comment:
Note: I went with `.*` here because checking for valid URL characters in regex is near unreadable. I'd be okay with going that route, however, just to tighten it up.

Reply:
I think this can be cleaned up a little, but haven't run through all the cases... If the `host` ends in `qnamaker/v5` or `qnamaker` and if `https://` is the prefix return `host`, otherwise return `'https://' + host`.

If someone provides [`'custom.some-other-site.tld/qnamaker'`](https://github.com/microsoft/botbuilder-js/pull/3559/files#diff-d52cfd00433cb19eb033fc54fb4e550d1caec9594e150b1b081bd82937c3767aR86), will they end up with `'https://custom.some-other-site.tld/qnamaker.azurewebsites.net/qnamaker'`? I'm not sure the `https://` check is necessary. Especially because someone could have their QnA Maker instance inside of an internal App Service Environment.

Reply:
There is a `url-parse` package we've used elsewhere that may help put an end to the RegExp madness.

Reply:
+1000 Let’s do that and just get out of this business.

Reply:
I'll see if I can integrate that in a way that isn't breaking (maybe somebody has something weird set up and `getHost()` makes it work for them?). Sorry for the late reply...looks like I need to adjust my email notifications

Reply:
@joshgummersall @stevengum Unfortunately, I don't know if `url-parse` is really the answer. I think the complexity is less due to regex and more due to trying to allow for any/all user input. 

It becomes something like this:

```typescript
    private getHost(dc: DialogContext): string {
        let value = this.hostname.getValue(dc.state);

        // Handle user including only host and not domain.
        if (!value.includes('.')) {
            value = value + '.azurewebsites.net';
        }

        // Handle user not including http(s). Needs to be set here or `url-parse` sets everything to pathname and not host.
        if (!/^https?:\/\//.test(value)) {
            value = 'https://' + value;
        }

        const url = new Url(value);

        // Handle *.azurewebsites.net domain
        if (url.host.endsWith('azurewebsites.net')) {
            url.set('protocol', 'https:');

            if (!url.pathname.includes('qnamaker/v5')) {
                // Default to /qnamaker, unless they specify the v5 API
                url.set('pathname', 'qnamaker');
            }
        } else {
            // Handle other domains, like one behind an Azure Application Gateway
            if (!url.protocol) {
                // Default to https, but allow http if set explicitly.
                url.set('protocol', 'https:');
            }

            // Custom domains may not use the same path, so we will not set the pathname here.
            // It is up to the developer to ensure they use an appropriate path name when using a custom domain.
        }

        return url.toString();
    }
```

Currently, this causes one of the tests to fail where the user inputs `https://myqnainstance.net/qnamaker.azurewebsites.net/qnamaker` and although I think the test is superfluous, I fear changing the code this much might break things for some users.

Thoughts?

*Sidenote: For v5, I'd like to do some checks for bad input and throw on launch. Force the user to use correct input, but let them know early and descriptively what they did wrong.*

Reply:
@joshgummersall @stevengum To add to that, I think this is better/cleaner:

```typescript
    private getHost(dc: DialogContext): string {
        let host = this.hostname.getValue(dc.state);

        // Handle no protocol.
        if (!/^https?:\/\//.test(host)) {
            host = 'https://' + host;
        }

        // Handle no domain.
        if (!host.includes('.')) {
            host = host + '.azurewebsites.net';
        }

        // Handle no pathname.
        if (!host.includes('qnamaker')) {
            host = host + '/qnamaker';
        }

        return host;
    }
```

The only issue is that it breaks [this test](https://github.com/microsoft/botbuilder-js/blob/main/libraries/botbuilder-ai/tests/qnaMakerDialog.test.js#L108), which I think is kind of an odd one.

Reply:
@stevengum thoughts on the test that breaks with the change above? Is `qnamaker.azurewebsites.net` really intended to be a supported `path` component?

Reply:
We can get rid of the negative test. Not sure why it's there... (I probably added it)

Going down the list:
1. It's probably reasonable to force https
2. Customers likely _aren't_ deploy a bot to their qnamaker app service, so checking for `'.'` seems reasonable
3. Seems reasonable

The cleaned up code looks good to me! 👍🏼 

**Side thought:**
`QnAMakerDialog.getHost()` isn't "complex" but it bothers me that we calculate the value constantly. Can we memoize it on a private StringExpression so that the SDK isn't constantly creating a usable `host`? I think this would be a language-specific feature.

Reply:
@stevengum It's a little difficult to read the diff, but I went with this:

```js
private getHost(dc: DialogContext): string {
        if (this.normalizedHostname) {
            return this.normalizedHostname;
        }

        let host = this.hostname.getValue(dc.state);

        // Handle no protocol.
        if (!/^https?:\/\//.test(host)) {
            host = 'https://' + host;
        }

        // Handle no domain.
        if (!host.includes('.')) {
            host = host + '.azurewebsites.net';
        }

        // Handle no pathname, only for azurewebsites.net domains.
        if (!host.includes('qnamaker') && !host.includes('azurewebsites.net')) {
            host = host + '/qnamaker';
        }

        this.normalizedHostname = host;

        return host;
    }
```

That work for memoization, or did you have something else in mind?

Reply:
Concerning memoization, isn't it important to preserve the fact that dialogs resolve these properties at runtime? I.e., can't the underlying state change, which might cause the normalized value to change? It might be a pedantic argument, but it seems like this could get us into trouble.

Reply:
resolved over Teams that memoization is good, but needed to allow for changes at runtime, which [this commit](https://github.com/microsoft/botbuilder-js/pull/3559/commits/0835c496b2a04936b1afbf9d602f59fcc7a2a87f) does